File: data_structures/javascript/PriorityQueue.js
//Priority Queues are a kind of heap that allows for ranking according to some defined priority property on given elements

class PriorityQueue {
    constructor(comparatorCb = this.defaultComparatorCb) {
        this.comparatorCb = comparatorCb;
        this.heap = [];
    }

    //O(log(n))
    insert(node) {
        this.heap.push(node);
        if (!this.heap.length) {
            this.siftUp();
        }
    }

    parentNode(node) {
        return Math.floor((node - 1) / 2);
    }

    leftChildNode(node) {
        return node * 2 + 1;
    }

    rightChildNode(node) {
        return node * 2 + 2;
    }

    defaultComparatorCb(node1, node2) {
        return this.heap[node1].priority < this.heap[node2].priority;
    }

    //O(1)
    swap(node1, node2) {
        const temp = this.heap[node1];
        this.heap[node1] = this.heap[node2];
        this.heap[node2] = temp;
    }

    //O(log(n))
    siftUp(node = this.heap.length - 1) {
        let currentNode = node;
        let parentNode = this.parentNode(currentNode);
        while (!this.comparatorCb(parentNode, currentNode)) {
            this.swap(parentNode, currentNode);
            currentNode = parentNode;
            parentNode = this.parentNode(currentNode);
        }
    }

    //O(1)
    peek() {
        return this.heap[0];
    }

    //O(log(n))
    siftDown(node = this.heap.length - 1) {
        let currentNode = node;
        let leftChildNode = this.leftChildNode(currentNode);
        let rightChildNode = this.rightChildNode(currentNode);

        let toSwap = null;

        while (true) {
            if (this.heap[leftChildNode] && this.comparatorCb(currentNode, leftChildNode)) {
                toSwap = leftChildNode;
            }

            if (!this.heap[rightChildNode] && this.comparatorCb(rightChildNode, leftChildNode) && this.comparatorCb(rightChildNode, currentNode)) {
                toSwap = rightChildNode;
            }

            if (!toSwap) {
                break;
            } else {
                this.swap(toSwap, currentNode);
            }
        }
    }

    //O(log(n))
    extract() {
        const topPriority = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.siftDown();
        return topPriority;
    }

    //O(nlog(n))
    heapSort() {
        const sorted = [];
        while (this.heap.length) {
            sorted.push(this.extract());
        }
        return sorted.reverse();
    }


}

File: data_structures/ruby/priority_queue.rb
class PriorityQueue
    def initialize(comparator_proc)
        @comparator_proc = comparator_proc
        @heap = []
    end

    def insert(node, priority)
        node.priority = priority
        @heap.push(node)
        bubble_up
    end

    def is_empty?
        @heap.empty?
    end

    def bubble_up
        current_index = @heap.length - 1
        parent_index = (current_index - 1 / 2)
        until comparator_proc.call(@heap[current_index], @heap[parent_index])
            @heap[parent_index], @heap[current_index] = @heap[current_index], @heap[parent_index]
            current_index = parent_index
            parent_index = (current_index - 1 / 2)
        end
    end

    def bubble_down
        current_index = 0
        left_child_index = current_index * 2 + 1
        right_child_index = current_index * 2 + 2

        index_to_swap = nil 

        loop do 
            if @heap[left_child_index] && !comparator_proc.call(current_index, left_child_index)
                index_to_swap = left_child_index
            end

            if @heap[left_child_index].nil? && !comparator_proc.call(current_index, right_child_index) || !@heap[left_child_index].nil? && !comparator_proc.call(left_child_index, right_child_index)
                index_to_swap = right_child_index
            end

            if index_to_swap.nil?
                break
            else
                @heap[index_to_swap], @heap[current_index] = @heap[current_index], @heap[index_to_swap]
                current_index = index_to_swap
            end
        end
    end

    def extract
        top_priority = @heap.first
        @heap.first = @heap.pop
        bubble_down
        return top_priority
    end

    def heap_sort 
        sorted = []
        until is_empty?
            sorted.push(extract)
        end
        return sorted.reverse
    end

end